File: snippets/dart_snippets/lib/receive_onchain.dart
import 'package:dart_snippets/sdk_instance.dart';
import 'package:breez_sdk/bridge_generated.dart';

Future<SwapInfo> generateReceiveOnchainAddress() async {
  // ANCHOR: generate-receive-onchain-address
  ReceiveOnchainRequest req = const ReceiveOnchainRequest();
  SwapInfo swapInfo = await breezSDK.receiveOnchain(req: req);

  // Send your funds to the below bitcoin address
  String address = swapInfo.bitcoinAddress;
  print(address);
  print("Minimum amount allowed to deposit in sats: ${swapInfo.minAllowedDeposit}");
  print("Maximum amount allowed to deposit in sats: ${swapInfo.maxAllowedDeposit}");
  return swapInfo;
  // ANCHOR_END: generate-receive-onchain-address
}

Future<SwapInfo?> getInProgressSwap() async {
  // ANCHOR: in-progress-swap
  SwapInfo? swapInfo = await breezSDK.inProgressSwap();
  print(swapInfo);
  // ANCHOR_END: in-progress-swap
  return swapInfo;
}

Future<List<SwapInfo>> listRefundables() async {
  // ANCHOR: list-refundables
  List<SwapInfo> refundables = await breezSDK.listRefundables();
  for (var refundable in refundables) {
    print(refundable.bitcoinAddress);
  }
  // ANCHOR_END: list-refundables
  return refundables;
}

Future<RefundResponse> executeRefund({
  required String swapAddress,
  required String toAddress,
  required int satPerVbyte,
}) async {
  // ANCHOR: execute-refund
  RefundRequest req = RefundRequest(
    swapAddress: swapAddress,
    toAddress: toAddress,
    satPerVbyte: satPerVbyte,
  );
  RefundResponse resp = await breezSDK.refund(req: req);
  print(resp.refundTxId);
  // ANCHOR_END: execute-refund
  return resp;
}

Future<OpenChannelFeeResponse> getChannelOpeningFees({
  required int amountMsat,
  int? expiry,
}) async {
  // ANCHOR: get-channel-opening-fees
  OpenChannelFeeRequest req = OpenChannelFeeRequest(amountMsat: amountMsat, expiry: expiry);
  OpenChannelFeeResponse resp = await breezSDK.openChannelFee(req: req);
  print(resp.feeMsat);
  // ANCHOR_END: get-channel-opening-fees
  return resp;
}

Future rescanSwaps() async {
  // ANCHOR: rescan-swaps
  await breezSDK.rescanSwaps();  
  // ANCHOR_END: rescan-swaps  
}

File: snippets/swift/BreezSDKExamples/Sources/ReceiveOnchain.swift
//
//  ReceiveOnchain.swift
//
//
//  Created by ruben on 14/11/2023.
//

import BreezSDK
import Foundation

func generateReceiveOnchainAddress(sdk: BlockingBreezServices) -> String? {
    // ANCHOR: generate-receive-onchain-address
    let swapInfo = try? sdk.receiveOnchain(req: ReceiveOnchainRequest())

    // Send your funds to the bellow bitcoin address
    let address = swapInfo?.bitcoinAddress
    print("Minimum amount allowed to deposit in sats: \(swapInfo!.minAllowedDeposit)")
    print("Maximum amount allowed to deposit in sats: \(swapInfo!.maxAllowedDeposit)")
    // ANCHOR_END: generate-receive-onchain-address

    return address
}

func getSwapInProgress(sdk: BlockingBreezServices) -> SwapInfo? {
    // ANCHOR: in-progress-swap
    let swapInfo = try? sdk.inProgressSwap()
    // ANCHOR_END: in-progress-swap
    return swapInfo
}

func listRefundables(sdk: BlockingBreezServices) -> [SwapInfo]? {
    // ANCHOR: list-refundables
    let refundables = try? sdk.listRefundables()
    // ANCHOR_END: list-refundables
    return refundables
}

func executeRefund(sdk: BlockingBreezServices, refundables: SwapInfo, satPerVbyte: UInt32) -> RefundResponse? {
    // ANCHOR: execute-refund
    let destinationAddress = "..."
    let response = try? sdk.refund(req: RefundRequest(swapAddress: refundables.bitcoinAddress, toAddress: destinationAddress, satPerVbyte: satPerVbyte))
    // ANCHOR_END: execute-refund
    return response
}

func getChannelOpeningFees(sdk: BlockingBreezServices, amountMsat: UInt64) -> OpenChannelFeeResponse? {
    // ANCHOR: get-channel-opening-fees
    let response = try? sdk.openChannelFee(req: OpenChannelFeeRequest(amountMsat: amountMsat))
    // ANCHOR_END: get-channel-opening-fees
    return response
}
